tex/generic/pgf/utilities/pgfutil-common.tex

```diff
@@ -49,6 +49,15 @@
 \def\pgfutil@trimspaces@@#1Q#2{#1}
 \catcode`\Q=11
 
+% \pgfutil@trimspaces@noexp{<token list>}
+%
+% Variant of \pgfutil@trimspaces that after full expansion leaves stripped
+% original text unexpanded in the input stream. Same as \trim@spaces@noexp
+% in trimspaces.sty
+\def\pgfutil@trimspaces@noexp#1{%
+  \unexpanded\expandafter\expandafter\expandafter{\pgfutil@trimspaces{#1}}%
```

Comment:
It's a pity that there's no `expl3` function to handle this two-step expansion case.

Reply:
`\exp_args:Nf` should do that. It basically equals ``\expandafter#1\expandafter{\romannumeral-`0#2}``

Reply:
We _used_ to have them (`d`-type), but there are very few places we found we needed them - as we've required e-TeX for a long time, we've been able to rely on most functions including `\unexpanded` to protect their return values. Also, now we have `\expanded`, we can use that ... sorry it doesn't work quite so well for you.

Reply:
What I have in my mind is sth more general (and/but less efficient, of course), like `\exp_by_steps:nn {<num of steps>} {<tokens>}`

```tex
\documentclass{article}
\usepackage{pgfmath}

\ExplSyntaxOn\makeatletter
% Expand #2 by exactly #1 steps and leaves the result in input stream. 
% The result is returned within \unexpanded.
\cs_new:Npn \exp_by_steps:nn #1#2
  {
    \int_compare:nNnTF {#1} > 0
      {
        \exp_by_steps:eo { \int_eval:n {#1-1} } {#2}
      }
      {
        \exp_not:n {#2}
      }
  }
\cs_generate_variant:Nn \exp_by_steps:nn {eo}

\exp_args:Ne \tl_show:n { \exp_by_steps:nn {0} {\undefined} }
\exp_args:Ne \tl_show:n { \exp_by_steps:nn {1} {\tl_if_eq:ccTF} }
\exp_args:Ne \tl_show:n { \exp_by_steps:nn {2} {\pgfutil@trimspaces{ \x}} }
\makeatother\ExplSyntaxOff

\begin{document}
\end{document}
```

```tex
% > \undefined .
% > \exp_args:Ncc \tl_if_eq:NNTF .
% > \x .
```